File: pixelpipes/image/filter.cpp

#include <opencv2/imgproc.hpp>

#include <pixelpipes/image.hpp>

/*
FILTER AND BLURING OPERATIONS
*/

namespace pixelpipes {

/**
 * @brief Blurs an image using a gaussian filter.
 * 
 */
SharedVariable GaussianBlur(std::vector<SharedVariable> inputs, ContextHandle context) noexcept(false) {

    VERIFY(inputs.size() == 5, "Incorrect number of parameters");

    cv::Mat image = Image::get_value(inputs[0]);
    int size_x = Integer::get_value(inputs[1]);
    int size_y = Integer::get_value(inputs[2]);
    float sigma_x = Float::get_value(inputs[3]);
    float sigma_y = Float::get_value(inputs[4]);

    cv::Mat result;
    cv::GaussianBlur(image, result, cv::Size(size_y, size_x), sigma_x, sigma_y, cv::BORDER_REPLICATE);

    return std::make_shared<Image>(result);
}

REGISTER_OPERATION_FUNCTION("gaussian_blur", GaussianBlur);

/**
 * @brief Blurs an image using a median filter.
 * 
 */
SharedVariable MedianBlur(std::vector<SharedVariable> inputs, ContextHandle context) noexcept(false) {

    VERIFY(inputs.size() == 2, "Incorrect number of parameters");

    cv::Mat image = Image::get_value(inputs[0]);
    int size = Integer::get_value(inputs[1]);

    cv::Mat result;
    cv::medianBlur(image, result, size);

    return std::make_shared<Image>(result);
}

REGISTER_OPERATION_FUNCTION("median_blur", MedianBlur);


/**
 * @brief Convolving an image with a normalized box filter.
 * 
 */
SharedVariable AverageBlur(std::vector<SharedVariable> inputs, ContextHandle context) noexcept(false) {

    VERIFY(inputs.size() == 2, "Incorrect number of parameters");

    cv::Mat image = Image::get_value(inputs[0]);
    int size = Integer::get_value(inputs[1]);

    cv::Mat result;
    cv::blur(image, result, cv::Size(size, size));

    return std::make_shared<Image>(result);
}

REGISTER_OPERATION_FUNCTION("average_blur", AverageBlur);

/**
 * @brief Applies the bilateral filter to an image.
 * 
 */
SharedVariable BilateralFilter(std::vector<SharedVariable> inputs, ContextHandle context) noexcept(false) {
    
    VERIFY(inputs.size() == 4, "Incorrect number of parameters");

    cv::Mat image = Image::get_value(inputs[0]);
    int d = Integer::get_value(inputs[1]);
    float sigma_color = Float::get_value(inputs[2]);
    float sigma_space = Float::get_value(inputs[3]);

    cv::Mat result;
    cv::bilateralFilter(image, result, d, (double)sigma_color, (double)sigma_space);

    return std::make_shared<Image>(result);
}

REGISTER_OPERATION_FUNCTION("bilateral_filter", BilateralFilter);


/**
 * @brief Convolves an image with custom kernel.
 * 
 */
SharedVariable LinearFilter(std::vector<SharedVariable> inputs, ContextHandle context) noexcept(false) {

    VERIFY(inputs.size() == 2, "Incorrect number of parameters");

    cv::Mat image = Image::get_value(inputs[0]);
    //cv::Mat kernel = Image::get_value(inputs[1]);
    // TEMP TEST FIX
    cv::Mat kernel = cv::Mat::ones(5, 5, CV_32F)/(float)(25);

    VERIFY(kernel.channels() == 1, "Kernel has more than one channel");

    cv::Mat result;
    cv::filter2D(image, result, image.depth(), kernel);

    return std::make_shared<Image>(result);
}

REGISTER_OPERATION_FUNCTION("linear_filter", LinearFilter);


}

File: pixelpipes/image/filter.py

from ..graph import EnumerationInput, Input, Macro, Operation
from .. import types
from ..image import BorderStrategy
from  .geometry import Transpose

class GaussianKernel(Operation):
    """Generate a Gaussian kernel
    """

    size = Input(types.Integer(), description="Size of the kernel, should be an odd integer")

    def operation(self):
        return "opencv:gaussian_kernel",

    def infer(self, size):
        return types.Image(depth="float")

class UniformKernel(Operation):
    """Generate a uniform kernel

    Inputs:
        - size: 

    Category: image, filters
    Tags: image, filter
    """

    size = Input(types.Integer(), description="Size of the kernel, should be an odd integer")

    def operation(self):
        return "opencv:uniform_kernel",

    def infer(self, size):
        return types.Image(depth="float")

class MedianBlur(Operation):
    """Blurs an image using a median filter."""

    source = Input(types.Image(), description="Input image")
    size = Input(types.Integer(), description="Size of the median window")
    
    def operation(self):
        return "opencv:median_blur",

    def infer(self, source, size):
        return types.Image(depth = source.element)

class BilateralFilter(Operation):
    """Applies the bilateral filter to an image.
    """

    source = Input(types.Image(), description="Input image")
    diameter = Input(types.Integer(), description="Diameter of pixel neighborhood")
    sigma_color = Input(types.Float(), description="Filter sigma in the color space")
    sigma_space = Input(types.Float(), description="Filter sigma in the coordinate space")
    
    def operation(self):
        return "opencv:bilateral_filter",

    def infer(self, **inputs):
        source = inputs["source"]
        return types.Image(depth = source.element)

class LinearFilter(Operation):
    """Linear filter

    Convolves an image with a custom kernel.

    Inputs:
        - source: source image
        - kernel: custom kernel

    Category: image, filters
    Tags: image, filter
    """

    source = Input(types.Image(), description="Input image")
    kernel = Input(types.Image(channels=1, depth="float"), description="Filter matrix")
    border = EnumerationInput(BorderStrategy, default="Reflect", description="Border handling strategy")

    def operation(self):
        return "opencv:linear_filter", 

    def infer(self,source, kernel, border):
        return types.Image(depth = source.element)

class GaussianFilter(Macro):
    """
    Blurs an image using a gaussian filter. Filtering is performed with two separate 1D filters.
    """

    source = Input(types.Image(), description="Input image")
    size_x = Input(types.Integer(), description="Filter size in X dimension")
    size_y = Input(types.Integer(), description="Filter size in Y dimension")
    border = EnumerationInput(BorderStrategy, default="Reflect", description="Border strategy")

    def expand(self, source, size_x, size_y, border):
        kernel_x = GaussianKernel(size_x)
        kernel_y = Transpose(GaussianKernel(size_y))
        tmp = LinearFilter(source, kernel_x, border=border)
        return LinearFilter(tmp, kernel_y, border=border)


class AverageFilter(Macro):
    """
    Convolving an image with a normalized box filter. Filtering is performed with two separate 1D filters.
    """

    source = Input(types.Image(), description="Input image")
    size_x = Input(types.Integer(), description="Filter size in X dimension")
    size_y = Input(types.Integer(), description="Filter size in Y dimension")
    border = EnumerationInput(BorderStrategy, default="Reflect", description="Border strategy")

    def expand(self, source, size_x, size_y, border):

        kernel_x = UniformKernel(size_x)
        kernel_y = Transpose(UniformKernel(size_y))
        tmp = LinearFilter(source, kernel_x, border=border)
        return LinearFilter(tmp, kernel_y, border=border)
